File: app/src/main/java/com/agrogames/islandsofwar/map/impl/Map.java
package com.agrogames.islandsofwar.map.impl;

import android.content.Context;
import android.util.Log;

import com.agrogames.islandsofwar.engine.abs.common.Cell;
import com.agrogames.islandsofwar.engine.abs.map.MapObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Map implements com.agrogames.islandsofwar.map.abs.Map {
    private final MapObject[] map;

    private Map(MapObject[] map) {
        this.map = map;
    }

    public static Map fromAssets(Context context, String name){
        try {
            return load(new Scanner(context.getAssets().open("maps/" + name)).useDelimiter("\\A").next());
        } catch (IOException e){
            Log.e("IOW", "Cannot load map with name " + name, e);
            return new Map(new MapObject[0]);
        }
    }

    private static Map load(String fileContent){
        List<MapObject> map = new ArrayList<>();
        String[] rows = fileContent.split("\n");
        for (int i = 0; i < rows.length; i++){
            String row = rows[i];
            for(int j = 0; j < row.length(); j++){
                char symbol = row.charAt(j);
                if(symbol == 'w'){
                    map.add(new Water(new Cell(j, 10-i)));
                }
            }
        }
        return new Map(map.toArray(new MapObject[0]));
    }

    @Override
    public MapObject[] getMap() {
        return map;
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/map/impl/Map.kt
package com.agrogames.islandsofwar.map.impl

import com.agrogames.islandsofwar.common.JsonBin
import com.agrogames.islandsofwar.engine.abs.common.Cell
import com.agrogames.islandsofwar.engine.abs.map.MapObject
import com.agrogames.islandsofwar.map.abs.MapParams
import okhttp3.OkHttpClient
import org.json.JSONObject
import java.util.Dictionary
import java.util.Hashtable

class Map private constructor(override val map: Array<MapObject>) : com.agrogames.islandsofwar.map.abs.Map {

    companion object {
        private val cache: Dictionary<String, String> = Hashtable()

        fun fromJsonBin(name: String): Map {
            if(name !in cache.keys().toList()){
                cache.put(name, JSONObject(OkHttpClient().newCall(JsonBin.maps).execute().body!!.string())
                    .getJSONObject("record").getString(name))
            }
            return load(cache.get(name))
        }

        private fun load(fileContent: String): Map {
            val map: MutableList<MapObject> = ArrayList()
            val rows = fileContent.split("\n".toRegex()).dropLastWhile { it.isEmpty() }
            for (i in rows.indices) {
                val row = rows[i]
                for (j in row.indices) {
                    val symbol = row[j]
                    if (symbol == 'w') {
                        map.add(Water(Cell(j, MapParams.Height - i)))
                    }
                }
            }
            return Map(map.toTypedArray())
        }
    }
}